opt: add rule to pull filters out of EXISTS condition

DrewKimball opened the pull request:
#### opt: add rule to pull filters out of EXISTS condition

This commit adds a new norm rule `HoistUnboundFilterFromExistsSubquery`
that pulls a filter out of an EXISTS filter if the inner filter only
references columns from the outer query. This gives other optimizations
a chance to apply before filter push-down rules move the EXISTS subquery
and everything in it below joins and other operators.

Fixes #146000

Release note: None

#### opt: add rule to pull filters out of EXISTS join condition

This commit adds a new norm rule `HoistUnboundJoinFilterFromExistsSubquery`
that pulls a filter from a join in an `EXISTS` subquery if the join filter
only references columns from the outer query. This gives other optimizations
a chance to apply before filter push-down rules move the EXISTS subquery
and everything in it below joins and other operators in the outer query.

Fixes #146000

Release note: None

#### opt: add session setting to control new norm rule

This commit adds a session setting `optimizer_use_exists_filter_hoist_rule`
to toggle the new optimizer rule `HoistUnboundFilterFromExistsSubquery`.
It if off by default in this commit, while the next will switch it on
only on master.

Informs #146000

Release note: None

#### opt: flip new session setting on

This commit sets `optimizer_use_exists_filter_hoist_rule` to true
by default. This commit will not be backported, so that the changes
in previous commits are opt-in.

Informs #146000

Release note: None

DrewKimball committed:
opt: add rule to pull filters out of EXISTS condition

This commit adds a new norm rule `HoistUnboundFilterFromExistsSubquery`
that pulls a filter out of an EXISTS filter if the inner filter only
references columns from the outer query. This gives other optimizations
a chance to apply before filter push-down rules move the EXISTS subquery
and everything in it below joins and other operators.

Fixes #146000

Release note: None

DrewKimball committed:
opt: add session setting to control new norm rule

This commit adds a session setting `optimizer_use_exists_filter_hoist_rule`
to toggle the new optimizer rule `HoistUnboundFilterFromExistsSubquery`.
It if off by default in this commit, while the next will switch it on
only on master.

Informs #146000

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146005)
<!-- Reviewable:end -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 10 of 10 files at r1, 8 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146005#-:-OPWgYWe-A6-RAKefBL1:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rytaft)

___
*[`pkg/sql/vars.go` line 4035 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146005#-OPWg5iv2bYMKHCq6skV:-OPWg5iv2bYMKHCq6skW:b-skp1m5) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb6223cb76cebcbcbbbc5a1a1fb7d1884cc7b19e/pkg/sql/vars.go#L4035)):*
> ```Go
> 			return formatBoolAsPostgresSetting(evalCtx.SessionData().OptimizerUseExistsFilterHoistRule), nil
> 		},
> 		GlobalDefault: globalTrue,
> ```

You can make backports easier if you default it to false first and then add a third commit to flip it to true. Then the first two commits should backport fairly cleanly. You'll have to update the optimizer tests to update enable the rule, but at least you won't have to re-do that for each backport.
___
*[`pkg/sql/opt/norm/testdata/rules/select` line 2482 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146005#-OPWdKb0BnLQP8RSpr6f:-OPWdKb0BnLQP8RSpr6g:b-daavhi) ([raw file](https://github.com/cockroachdb/cockroach/blob/8911c7f282bb3a2ed39b92179924ccd55f5469b1/pkg/sql/opt/norm/testdata/rules/select#L2482)):*
> ```
> 
> # --------------------------------------------------
> # HoistUnboundFilterFromExistsSubquery
> ```

It would be interesting to see a case where the subquery is a join with an ON condition that references only outer columns. Maybe this rule will somehow handle that? Or it will reveal that we might need another rule for that case.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Could the same hoisting be done for other subqueries with filters? For example, = ANY() or IN()?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146005#-:-OPXSt_W6pmiJPfQNhnG:b-4o3uip)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @rytaft)



<!-- Sent from Reviewable.io -->

DrewKimball committed:
opt: flip new session setting on

This commit sets `optimizer_use_exists_filter_hoist_rule` to true
by default. This commit will not be backported, so that the changes
in previous commits are opt-in.

Informs #146000

Release note: None

DrewKimball committed:
opt: add rule to pull filters out of EXISTS join condition

This commit adds a new norm rule `HoistUnboundJoinFilterFromExistsSubquery`
that pulls a filter from a join in an `EXISTS` subquery if the join filter
only references columns from the outer query. This gives other optimizations
a chance to apply before filter push-down rules move the EXISTS subquery
and everything in it below joins and other operators in the outer query.

Fixes #146000

Release note: None

DrewKimball reviewed (commented):
>Could the same hoisting be done for other subqueries with filters? For example, = ANY() or IN()?

It can - they get normalized to EXISTS, so are already handled. I had an `ANY` case already, and just added an `IN` case as well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146005#-:-OPYQiJiA3HMP-Pkeay0:b-usj9u3)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[`pkg/sql/vars.go` line 4035 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146005#-OPWg5iv2bYMKHCq6skV:-OPYQsnM8f_xVoGaF7vr:b1gsn0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb6223cb76cebcbcbbbc5a1a1fb7d1884cc7b19e/pkg/sql/vars.go#L4035)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

You can make backports easier if you default it to false first and then add a third commit to flip it to true. Then the first two commits should backport fairly cleanly. You'll have to update the optimizer tests to update enable the rule, but at least you won't have to re-do that for each backport.
</blockquote></details>

Great idea - Done.
___
*[`pkg/sql/opt/norm/testdata/rules/select` line 2482 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146005#-OPWdKb0BnLQP8RSpr6f:-OPZ-d_R4XchiptJbpq2:blvvpge) ([raw file](https://github.com/cockroachdb/cockroach/blob/8911c7f282bb3a2ed39b92179924ccd55f5469b1/pkg/sql/opt/norm/testdata/rules/select#L2482)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It would be interesting to see a case where the subquery is a join with an ON condition that references only outer columns. Maybe this rule will somehow handle that? Or it will reveal that we might need another rule for that case.
</blockquote></details>

Good point - I added another commit with a rule that handles inner-joins and semi-joins.


<!-- Sent from Reviewable.io -->

mgartner commented on pkg/sql/opt/norm/testdata/rules/decorrelate:
```diff
@@ -8132,3 +8132,646 @@ project
+      └── filters
+           └── x:8 = k:1 [outer=(1,8), constraints=(/1: (/NULL - ]; /8: (/NULL - ]), fd=(1)==(8), (8)==(1)]
+
+norm expect=HoistUnboundJoinFilterFromExistsSubquery
```
nit: Is this test supposed to be in the section below?

DrewKimball replied:
Yes, Done.

mgartner reviewed (approved):
LGTM. Don't forget to update the PR description with the new commit messages.

DrewKimball commented:
bors r=mgartner
